File: gems/decomplex/lib/decomplex/weighted_inlined_cognitive_complexity.rb
# frozen_string_literal: true

require "set"
require_relative "syntax"
require_relative "structural_topology"

module Decomplex
  # Scores the reader burden of a method after conservatively inlining
  # same-owner bare/self helper calls. This catches "small" orchestration
  # methods whose complexity was moved into private/single-use helpers.
  class WeightedInlinedCognitiveComplexity
    MethodBody = Struct.new(:id, :owner, :name, :file, :line, :span, :node,
                            :complexity, keyword_init: true)
    LocalScore = Struct.new(:id, :owner, :name, :file, :line, :span, :score, :signals, keyword_init: true)
    Contribution = Struct.new(:callee_id, :callee_name, :score, :weight, :depth, :chain, keyword_init: true)

    DEFAULT_MIN_SCORE = 12.0
    DEFAULT_MIN_HIDDEN = 15.0
    DEFAULT_MAX_DEPTH = 2
    DEPTH_WEIGHTS = [1.0, 1.0, 0.6, 0.35].freeze
    EDGE_WEIGHTS = { always: 1.0, conditional: 0.75, iterates: 1.15 }.freeze
    def self.scan(files, min_score: DEFAULT_MIN_SCORE, min_hidden: DEFAULT_MIN_HIDDEN, max_depth: DEFAULT_MAX_DEPTH)
      new(files, min_score: min_score, min_hidden: min_hidden, max_depth: max_depth).scan
    end

    def initialize(files, min_score:, min_hidden:, max_depth:)
      @files = files
      @min_score = min_score.to_f
      @min_hidden = min_hidden.to_f
      @max_depth = max_depth.to_i
    end

    def scan
      topology = StructuralTopology.scan(@files)
      bodies = syntax_method_bodies
      scores = bodies.to_h do |body|
        score = body.complexity
        [body.id, LocalScore.new(
          id: body.id,
          owner: body.owner,
          name: body.name,
          file: body.file,
          line: body.line,
          span: body.span,
          score: score[:score],
          signals: score[:signals]
        )]
      end

      Analyzer.new(topology, scores, @min_score, @min_hidden, @max_depth).findings
    end

    private

    def syntax_method_bodies
      @files.flat_map do |file|
        document = Syntax.parse(file, parser: "tree_sitter")
        score_by_id = document.local_complexity_scores
        document.local_methods.map do |method|
          method_body(method, complexity: score_by_id.fetch(method.id, { score: 0.0, signals: {} }))
        end
      end
    end

    def method_body(summary, complexity:)
      owner = summary.owner == "(top-level)" ? "(top-level:#{summary.file})" : summary.owner
      MethodBody.new(
        id: "#{owner}##{summary.name}",
        owner: owner,
        name: summary.name,
        file: summary.file,
        line: summary.line,
        span: summary.span,
        node: summary.node,
        complexity: complexity
      )
    end

    class Analyzer
      def initialize(topology, scores, min_score, min_hidden, max_depth)
        @topology = topology
        @scores = scores
        @min_score = min_score
        @min_hidden = min_hidden
        @max_depth = max_depth
      end

      def findings
        @scores.values.filter_map { |score| finding_for(score) }
               .sort_by { |row| [-row[:hidden], -row[:inlined], row[:file].to_s, row[:method].to_s] }
      end

      private

      def finding_for(score)
        contributions = inlined_contributions(score.id, depth: 1, visited: Set[score.id])
        hidden = round(contributions.sum(&:score))
        total = round(score.score + hidden)
        return nil if total < @min_score || hidden < @min_hidden

        direct_single_caller = single_caller_callees(score.id)
        at = "#{score.file}:#{score.name}:#{score.line}"
        {
          at: at,
          owner: score.owner,
          method: score.name,
          local: score.score,
          inlined: total,
          hidden: hidden,
          depth: contributions.map(&:depth).max || 0,
          single_caller_callees: direct_single_caller,
          call_chain: strongest_chain(score, contributions),
          reason: reason(score, hidden, direct_single_caller),
          signals: score.signals,
          spans: { at => score.span }
        }
      end

      def inlined_contributions(method_id, depth:, visited:)
        return [] if depth > @max_depth

        grouped_edges(method_id).flat_map do |edge|
          next [] if visited.include?(edge.callee)

          callee = @scores[edge.callee]
          next [] unless callee

          weight = contribution_weight(edge, depth)
          direct = Contribution.new(
            callee_id: edge.callee,
            callee_name: edge.callee_name,
            score: round(callee.score * weight),
            weight: round(weight),
            depth: depth,
            chain: [edge.callee_name]
          )
          nested = inlined_contributions(edge.callee, depth: depth + 1, visited: visited + [edge.callee])
          nested = nested.map do |contribution|
            Contribution.new(
              callee_id: contribution.callee_id,
              callee_name: contribution.callee_name,
              score: round(contribution.score * weight),
              weight: round(contribution.weight * weight),
              depth: contribution.depth,
              chain: [edge.callee_name] + contribution.chain
            )
          end
          [direct] + nested
        end
      end

      def grouped_edges(method_id)
        @topology.internal_calls(method_id)
                 .group_by(&:callee)
                 .values
                 .map { |edges| edges.max_by { |edge| EDGE_WEIGHTS.fetch(edge.type, 1.0) } }
      end

      def contribution_weight(edge, depth)
        caller_factor = @topology.single_internal_caller?(edge.callee) ? 1.0 : 0.35
        visibility_factor = shared_public_step?(edge) ? 0.6 : 1.0
        depth_factor = DEPTH_WEIGHTS.fetch(depth, DEPTH_WEIGHTS.last)
        edge_factor = EDGE_WEIGHTS.fetch(edge.type, 1.0)
        caller_factor * visibility_factor * depth_factor * edge_factor
      end

      def shared_public_step?(edge)
        @topology.visibility(edge.callee) == :public && !@topology.single_internal_caller?(edge.callee)
      end

      def single_caller_callees(method_id)
        grouped_edges(method_id).filter_map do |edge|
          edge.callee_name if @topology.single_internal_caller?(edge.callee)
        end.sort
      end

      def strongest_chain(score, contributions)
        chain = contributions.max_by(&:score)&.chain || []
        [score.name] + chain
      end

      def reason(score, hidden, single_caller_callees)
        if single_caller_callees.empty?
          "same-owner call chain adds #{hidden} weighted cognitive points"
        else
          "#{single_caller_callees.size} single-caller helper(s) add #{hidden} weighted cognitive points"
        end
      end

      def round(value)
        (value * 10).round / 10.0
      end
    end
  end
end

File: gems/decomplex/rust/src/decomplex/detectors/weighted_inlined_cognitive_complexity.rs
use crate::decomplex::ast::Span;
use crate::decomplex::detectors::{local_flow, structural_topology};
use crate::decomplex::syntax::{self, Document, Language, LocalComplexityScore};
use anyhow::Result;
use serde::Serialize;
use std::collections::{BTreeMap, BTreeSet};
use std::path::PathBuf;

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct WeightedInlinedCognitiveComplexityRow {
    pub at: String,
    pub owner: String,
    pub method: String,
    pub local: f64,
    pub inlined: f64,
    pub hidden: f64,
    pub depth: usize,
    pub single_caller_callees: Vec<String>,
    pub call_chain: Vec<String>,
    pub reason: String,
    pub signals: BTreeMap<String, usize>,
    pub spans: BTreeMap<String, Span>,
}

pub fn scan_files(
    files: &[PathBuf],
    language: Language,
) -> Result<Vec<WeightedInlinedCognitiveComplexityRow>> {
    let documents = syntax::parse_files(files, language)?;
    Ok(scan_documents(&documents))
}

pub fn scan_documents(documents: &[Document]) -> Vec<WeightedInlinedCognitiveComplexityRow> {
    let summaries = local_flow::scan_documents(documents);
    scan_documents_with_summaries(documents, &summaries)
}

pub fn scan_documents_with_summaries(
    documents: &[Document],
    summaries: &[local_flow::MethodSummary],
) -> Vec<WeightedInlinedCognitiveComplexityRow> {
    let topology_report = structural_topology::scan_documents(documents);
    let topology = structural_topology::Graph::new(topology_report.methods, topology_report.edges);
    let complexity_scores = documents
        .iter()
        .flat_map(|document| {
            document
                .local_complexity_scores
                .iter()
                .map(|(id, score)| ((document.file.clone(), id.clone()), score.clone()))
        })
        .collect::<BTreeMap<_, _>>();

    let mut scores = BTreeMap::new();
    for summary in summaries {
        let owner = if summary.owner == "(top-level)" {
            format!("(top-level:{})", summary.file)
        } else {
            summary.owner.clone()
        };
        let id = format!("{}#{}", owner, summary.name);
        let score = complexity_scores
            .get(&(summary.file.clone(), summary.id.clone()))
            .cloned()
            .unwrap_or_else(|| LocalComplexityScore {
                score: 0.0,
                signals: BTreeMap::new(),
            });
        scores.insert(
            id.clone(),
            LocalScore {
                id,
                owner,
                name: summary.name.clone(),
                file: summary.file.clone(),
                line: summary.line,
                span: summary.span,
                score: score.score,
                signals: score.signals,
            },
        );
    }

    let analyzer = Analyzer::new(topology, scores, 12.0, 15.0, 2);
    analyzer.findings()
}

struct LocalScore {
    id: String,
    owner: String,
    name: String,
    file: String,
    line: usize,
    span: Span,
    score: f64,
    signals: BTreeMap<String, usize>,
}

struct Contribution {
    #[allow(dead_code)]
    callee_id: String,
    callee_name: String,
    score: f64,
    #[allow(dead_code)]
    weight: f64,
    depth: usize,
    chain: Vec<String>,
}

fn format_one_decimal(value: f64) -> String {
    format!("{value:.1}")
}

struct Analyzer {
    topology: structural_topology::Graph,
    scores: BTreeMap<String, LocalScore>,
    min_score: f64,
    min_hidden: f64,
    max_depth: usize,
}

impl Analyzer {
    fn new(
        topology: structural_topology::Graph,
        scores: BTreeMap<String, LocalScore>,
        min_score: f64,
        min_hidden: f64,
        max_depth: usize,
    ) -> Self {
        Self {
            topology,
            scores,
            min_score,
            min_hidden,
            max_depth,
        }
    }

    fn findings(&self) -> Vec<WeightedInlinedCognitiveComplexityRow> {
        let mut out: Vec<_> = self
            .scores
            .values()
            .filter_map(|s| self.finding_for(s))
            .collect();
        out.sort_by(|a, b| {
            b.hidden
                .partial_cmp(&a.hidden)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| {
                    b.inlined
                        .partial_cmp(&a.inlined)
                        .unwrap_or(std::cmp::Ordering::Equal)
                })
                .then_with(|| a.at.cmp(&b.at))
        });
        out
    }

    fn finding_for(&self, score: &LocalScore) -> Option<WeightedInlinedCognitiveComplexityRow> {
        let mut visited = BTreeSet::new();
        visited.insert(score.id.clone());
        let contributions = self.inlined_contributions(&score.id, 1, &mut visited);

        let hidden = self.round(contributions.iter().map(|c| c.score).sum());
        let total = self.round(score.score + hidden);
        if total < self.min_score || hidden < self.min_hidden {
            return None;
        }

        let direct_single_caller = self.single_caller_callees(&score.id);
        let at = format!("{}:{}:{}", score.file, score.name, score.line);
        let mut spans = BTreeMap::new();
        spans.insert(at.clone(), score.span);

        Some(WeightedInlinedCognitiveComplexityRow {
            at,
            owner: score.owner.clone(),
            method: score.name.clone(),
            local: score.score,
            inlined: total,
            hidden,
            depth: contributions.iter().map(|c| c.depth).max().unwrap_or(0),
            single_caller_callees: direct_single_caller.clone(),
            call_chain: self.strongest_chain(score, &contributions),
            reason: self.reason(hidden, &direct_single_caller),
            signals: score.signals.clone(),
            spans,
        })
    }

    fn inlined_contributions(
        &self,
        method_id: &str,
        depth: usize,
        visited: &mut BTreeSet<String>,
    ) -> Vec<Contribution> {
        if depth > self.max_depth {
            return Vec::new();
        }

        let mut out = Vec::new();
        for edge in self.grouped_edges(method_id) {
            if visited.contains(&edge.callee) {
                continue;
            }
            let Some(callee) = self.scores.get(&edge.callee) else {
                continue;
            };

            let weight = self.contribution_weight(&edge, depth);
            let direct = Contribution {
                callee_id: edge.callee.clone(),
                callee_name: edge.callee_name.clone(),
                score: self.round(callee.score * weight),
                weight: self.round(weight),
                depth,
                chain: vec![edge.callee_name.clone()],
            };

            let mut next_visited = visited.clone();
            next_visited.insert(edge.callee.clone());
            let nested = self.inlined_contributions(&edge.callee, depth + 1, &mut next_visited);
            let nested: Vec<_> = nested
                .into_iter()
                .map(|c| Contribution {
                    callee_id: c.callee_id,
                    callee_name: c.callee_name,
                    score: self.round(c.score * weight),
                    weight: self.round(c.weight * weight),
                    depth: c.depth,
                    chain: {
                        let mut chain = vec![edge.callee_name.clone()];
                        chain.extend(c.chain);
                        chain
                    },
                })
                .collect();

            out.push(direct);
            out.extend(nested);
        }
        out
    }

    fn grouped_edges(&self, method_id: &str) -> Vec<structural_topology::Edge> {
        let mut by_callee: BTreeMap<String, Vec<structural_topology::Edge>> = BTreeMap::new();
        for edge in self.topology.internal_calls(method_id) {
            by_callee.entry(edge.callee.clone()).or_default().push(edge);
        }
        by_callee
            .into_iter()
            .map(|(_, edges)| {
                edges
                    .into_iter()
                    .fold(None, |best: Option<structural_topology::Edge>, edge| {
                        let Some(current) = best else {
                            return Some(edge);
                        };
                        if self.edge_weight(&edge.r#type) > self.edge_weight(&current.r#type) {
                            Some(edge)
                        } else {
                            Some(current)
                        }
                    })
                    .unwrap()
            })
            .collect()
    }

    fn contribution_weight(&self, edge: &structural_topology::Edge, depth: usize) -> f64 {
        let caller_factor = if self.topology.single_internal_caller(&edge.callee) {
            1.0
        } else {
            0.35
        };
        let visibility_factor = if self.shared_public_step(edge) {
            0.6
        } else {
            1.0
        };
        let depth_factor = match depth {
            1 => 1.0,
            2 => 0.6,
            _ => 0.35,
        };
        let edge_factor = self.edge_weight(&edge.r#type);
        caller_factor * visibility_factor * depth_factor * edge_factor
    }

    fn edge_weight(&self, t: &str) -> f64 {
        match t {
            "always" => 1.0,
            "conditional" => 0.75,
            "iterates" => 1.15,
            _ => 1.0,
        }
    }

    fn shared_public_step(&self, edge: &structural_topology::Edge) -> bool {
        self.topology.visibility(&edge.callee) == Some("public")
            && !self.topology.single_internal_caller(&edge.callee)
    }

    fn single_caller_callees(&self, method_id: &str) -> Vec<String> {
        let mut out: Vec<_> = self
            .grouped_edges(method_id)
            .into_iter()
            .filter(|e| self.topology.single_internal_caller(&e.callee))
            .map(|e| e.callee_name)
            .collect();
        out.sort();
        out
    }

    fn strongest_chain(&self, score: &LocalScore, contributions: &[Contribution]) -> Vec<String> {
        let chain = contributions
            .iter()
            .fold(None, |best: Option<&Contribution>, contribution| {
                let Some(current) = best else {
                    return Some(contribution);
                };
                if contribution.score > current.score {
                    Some(contribution)
                } else {
                    Some(current)
                }
            })
            .map(|c| c.chain.clone())
            .unwrap_or_default();
        let mut out = vec![score.name.clone()];
        out.extend(chain);
        out
    }

    fn reason(&self, hidden: f64, single_caller_callees: &[String]) -> String {
        if single_caller_callees.is_empty() {
            format!(
                "same-owner call chain adds {} weighted cognitive points",
                format_one_decimal(hidden)
            )
        } else {
            format!(
                "{} single-caller helper(s) add {} weighted cognitive points",
                single_caller_callees.len(),
                format_one_decimal(hidden)
            )
        }
    }

    fn round(&self, value: f64) -> f64 {
        (value * 10.0).round() / 10.0
    }
}
